feat(components): Add `ConcurrentGroup` and friends

SyntaxColoring opened the pull request:
## Overview

Implements new components for Protocol Designer: `ConcurrentGroup`, `ConcurrentGroupCheckpoint` and `ConcurrentGroupStepContainer`. Closes EXEC-1686. See there for a link to the designs.

This is currently based atop #19298. It will need to merge into `edge` after it.

## Test Plan and Hands on Testing

Preview: https://s3-us-west-2.amazonaws.com/opentrons-components/new_component_checkpoints/index.html?path=/docs/protocol-designer-molecules-concurrentgroup

## Changelog

* Implement the new component, as described in EXEC-1686.
* Address an problem in `Chip` where the icon would shrink when text got long.

## Review requests

See comments.

## Risk assessment

None. Unused so far.

SyntaxColoring committed:
WIP: Checkpoint and CheckpointList

SyntaxColoring committed:
Fix up an incomplete rename.

SyntaxColoring committed:
Replace flexbox gap with a per-step internal padding.

SyntaxColoring committed:
Comment.

SyntaxColoring committed:
Fix spacing around drag-n-drop divider.

SyntaxColoring committed:
Update unit tests to be able to find the style on the right element.

SyntaxColoring committed:
Fix outdated unit test filename.

SyntaxColoring committed:
Fix <Chip> icon shrinking when the <Chip> is squeezed and its text wraps.

SyntaxColoring committed:
Lint + format.

SyntaxColoring commented on components/src/atoms/Chip/index.tsx:
```diff
@@ -156,6 +156,7 @@ const ICON_STYLE = (
 ): FlattenSimpleInterpolation => css`
   width: ${chipSize === 'medium' ? '1rem' : smallSize};
   height: ${chipSize === 'medium' ? '1rem' : smallSize};
+  flex-shrink: 0;
```
This fixes the `Chip` icon slowly shrinking when the text gets long enough to wrap. I guess this use of `Chip` is the first time this has happened in practice.

[Edit] `Chip` is no longer used in this PR, but this still seems like a good idea to fix, right?

SyntaxColoring committed:
Delete unnecessary clsx.

SyntaxColoring committed:
Grammar.

SyntaxColoring committed:
Format CSS.

SyntaxColoring committed:
Better name.

SyntaxColoring commented on components/src/molecules/Checkpoints/checkpoint.module.css:
```diff
@@ -0,0 +1,29 @@
+@import '../../styles/global.css';
```
This was a hack to get CSS modules to work in Storybook. Now that https://github.com/Opentrons/opentrons/pull/19312 is merged, I'll delete this before merging.

SyntaxColoring commented on components/src/molecules/Checkpoints/commonspacing.module.css:
```diff
@@ -0,0 +1,16 @@
+/*
+Styles to add content-dependent spacing between children within a Checkpoint.
+
+- CheckpointStepContainers in a contiguous sequence have no space between them.
+- CheckpointChips in a contiguous sequence have --spacing-8 between them.
+- Between a CheckpointStepContainer and CheckpointChip is --spacing-8.
+
+An alternative method would have been to use JS to find contiguous sequences and put a
+wrapper component around each one, but that would interfere with <ul> nesting semantics.
+*/
+
+.chip + .chip,
+.chip + .step_container,
+.step_container + .chip {
+  margin-top: var(--spacing-8);
+}
```
Does this approach seem OK?

koji replied:
this would depend on the team's preference.
i'm open to any approach . if we decide to go with this approach, I think it's important to make sure the whole team understands how it's meant to be used.

koji committed:
fix(monorepo): import global css to fix css loading issue in storybook (#19312)

* fix(monorepo): import global css to fix css loading issue in storybook

(cherry picked from commit f89067f38afa6dfdd6b2d6116b86b454451037d8)

koji commented on components/src/molecules/Checkpoints/Checkpoint.tsx:
this component is for PD so this should be under `protocol-designer/src/components/molecules`.

SyntaxColoring replied:
Wow nice! I didn't think it would show up in Storybook if I added it there, but it does.

koji commented on components/src/molecules/Checkpoints/checkpoint.module.css:
```diff
@@ -0,0 +1,29 @@
+  flex-grow: 0;
+  flex-shrink: 0;
+  border-radius: calc(var(--checkpointlist-line-stroke-width) / 2);
+  margin: var(--spacing-4);
```
is it possible to avoid using margin?

SyntaxColoring replied:
Sure, it would be a nested div or `Box` like this:

```html
<div> <!-- Just for spacing. Contains `padding: var(--spacing-4)`. -->
    <div></div> <!-- The actual line. No padding or spacing. Has `width`, `border-radius`, and `background-color`. -->
</div>
```

I can do that if you think it's cleaner.

koji commented on components/src/molecules/Checkpoints/Checkpoint.tsx:
```diff
@@ -0,0 +1,32 @@
+export function Checkpoint(props: PropsWithChildren<Props>): JSX.Element {
+  const { active, children } = props
+  return (
+    <div className={clsx(styles.container, { [styles.active]: active })}>
```
you can use Flex since i'm refactoring Flex to keep the current implementation with CSS Modules.

koji commented:
for bullet-point



<img width="151" height="214" alt="Screenshot 2025-08-21 at 3 58 29 PM" src="https://github.com/user-attachments/assets/85f8310e-b1e1-4c20-b93a-08cea587f2b4" />

koji commented on components/src/molecules/Checkpoints/CheckpointChip.tsx:
```diff
@@ -0,0 +1,25 @@
+  const { text } = props
+  return (
+    <div className={styles.chip}>
+      <Chip
```
as i posted, the icon usage of Chip - `connection-status` is different from this component.

SyntaxColoring committed:
Move Checkpoint and StepContainer to protocol-designer/.

I put these in components/ because I thought that was the only way to get them to show up in Storybook, but it turns out that Storybook picks them up just fine if they're kept local to protocol-designer.

SyntaxColoring committed:
Delete my hack for getting Storybook to load global CSS.

koji commented:
@SyntaxColoring 
number of checkpoints: no limit in Figma.
do you have any ideas of the max number? i'm asking this because if there are so many checkpoints, maybe we need to handle overflow in this component.

SyntaxColoring committed:
Design update: --spacing-4 around chips, not --spacing-8.

SyntaxColoring committed:
Replace <Chip> with a locally defined bullet point.

SyntaxColoring committed:
Rephrase comment for readability.

SyntaxColoring committed:
Use <Flex> and <Box> instead of CSS modules.

SyntaxColoring commented on protocol-designer/src/components/molecules/Checkpoint/CheckpointChip.tsx:
```diff
@@ -0,0 +1,34 @@
+  text: string
+}
+
+/** A non-interactable, system-provided step within a `Checkpoint`. */
+export function CheckpointChip(props: Props): JSX.Element {
```
This needs a new name now that it's no longer implemented with `Chip`. I'm waiting to see how the discussion in Figma turns out.

SyntaxColoring committed:
Replace margin with padding.

mjhuff commented on protocol-designer/src/components/molecules/Checkpoint/Checkpoint.tsx:
```diff
@@ -0,0 +1,56 @@
+
+import type { PropsWithChildren } from 'react'
+
+interface Props {
```
```suggestion
interface CheckpointProps {
```
Would prefer this just for code consistency and not needing to cast imports when this interface is imported elsewhere.

SyntaxColoring replied:
Done in 87d3a2ddf9d35f921b13ca1b2d0e428456150a7c, but that's kind of surprising to me. You prefer that over `React.ComponentProps<Checkpoint>`? Seems like `React.ComponentProps` would pollute the global namespace less.

mjhuff replied:
I don't prefer it, but it's the pattern we follow in our codebase, and IMO consistency is more important on this one.

mjhuff commented on protocol-designer/src/components/molecules/Checkpoint/Checkpoint.tsx:
```diff
@@ -0,0 +1,56 @@
+import { Box, COLORS, Flex, SPACING } from '@opentrons/components'
+
+import type { PropsWithChildren } from 'react'
```
Very nice find

mjhuff commented on protocol-designer/src/components/molecules/Checkpoint/CheckpointChip.tsx:
```diff
@@ -0,0 +1,34 @@
+
+import spacingStyles from './commonspacing.module.css'
+
+interface Props {
```
```suggestion
interface WhateverDesignNamesThisProps {
```
ditto

mjhuff commented on protocol-designer/src/components/molecules/Checkpoint/CheckpointStepContainer.tsx:
```diff
@@ -0,0 +1,18 @@
+  props: ComponentProps<typeof StepContainer>
+): JSX.Element {
+  return (
+    <div className={spacingStyles.step_container}>
```
Any reason we aren't using `Flex` or `Box `here?

SyntaxColoring replied:
Yeah. `spacingStyles.step_container` uses CSS adjacent sibling selectors that can't be written in terms of `Flex`/`Box` style props. I could use `Box` without any style props, but at that point, what's the point of `Box`?

koji replied:
i think using `Box` would be good since I'm refactoring all primitives (Flex, Box, Link, Text, Btn, and Svg) to support our current implementation.
the point is consistency of our code base. i think we use div for a really special case.

mjhuff commented on protocol-designer/src/components/molecules/Checkpoint/CheckpointStepContainer.tsx:
```diff
@@ -0,0 +1,18 @@
+import { StepContainer } from '../StepContainer'
+import spacingStyles from './commonspacing.module.css'
```
We should ticket out adding .module.css to our import order linting. Feel free to assign this to me if you'd like!

SyntaxColoring replied:
EXEC-1771. TY!

mjhuff reviewed (approved):
Nice job! I think @koji has covered most of the feedback already (thank you Koji). Just a few clean up comments/Qs.

koji commented:
Could you add tests for the new components you made?

SyntaxColoring committed:
Add todo comments for naming, and rename Props types.

koji commented on protocol-designer/src/components/molecules/Checkpoint/Checkpoint.tsx:
```diff
@@ -0,0 +1,61 @@
+  const { active, children } = props
+  return (
+    <Flex
+      flexDirection="row"
```
```suggestion
      flexDirection={DIRECTION_ROW}
```

koji commented on protocol-designer/src/components/molecules/Checkpoint/Checkpoint.tsx:
```diff
@@ -0,0 +1,61 @@
+      padding={SPACING.spacing2}
+    >
+      <Box
+        width={SPACING.spacing4}
```
we use SPACING for padding, gap, margin etc.
for height, width, and size, we use the fixed value.

```suggestion
        width="0.25rem"
```

SyntaxColoring replied:
Thank you, I was wondering about this!

koji commented on protocol-designer/src/components/molecules/Checkpoint/CheckpointChip.tsx:
```diff
@@ -0,0 +1,37 @@
+
+function Bullet(): JSX.Element {
+  return (
+    <svg width={SPACING.spacing4} height={SPACING.spacing4} viewBox="-1 -1 2 2">
```
```suggestion
    <svg width="0.25rem" height="0.25rem" viewBox="-1 -1 2 2">
```

SyntaxColoring committed:
Use plain units, not spacing tokens, for width and height.

Co-authored-by: koji <baxin1919@gmail.com>

SyntaxColoring committed:
Oops, if the parent is a <ul>, then the children need to be <li>s.

SyntaxColoring committed:
Many renames.

The new hierarchy:

* ConcurrentGroup (renamed from Checkpoint)
  * ConcurrentGroupCheckpoint (renamed from CheckpointChip)
  * ConcurrentGroupStepContainer (renamed from CheckpointStepContainer)

SyntaxColoring committed:
Add unit tests.

SyntaxColoring committed:
Use the DIRECTION_ROW constant.

Co-authored-by: koji <baxin1919@gmail.com>

SyntaxColoring committed:
Post-rename fixups.

koji reviewed (approved):
nice work!
lgtm

whether you wait for Mel’s response to the spacing question and include it in this PR, or follow up in a separate one. that’s up to you.

SyntaxColoring committed:
Merge branch 'edge' into gap_via_padding

Resolve conflicts in StepContainer.tsx.

SyntaxColoring committed:
Merge branch 'gap_via_padding' into new_component_checkpoints

Catching this branch up to the latest changes in its original base branch, which has now merged into edge.

SyntaxColoring committed:
Merge branch 'edge' into new_component_checkpoints

Resolve conflicts in protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/ConnectedStepContainer.tsx.

SyntaxColoring committed:
Change spacing between children from 4 to 8.

This was formerly inconsistent in Figma. Figma's now been fixed to clarify that 8 is the intent.

SyntaxColoring committed:
Fix grey colors.

The text should be grey60. I think it was accidentally defaulting to grey50?

Also, the bullet should match the text color.
